File: ruby/lib/runapi/runway_aleph/resources/video_to_video.rb
# frozen_string_literal: true

module RunApi
  module RunwayAleph
    module Resources
      class VideoToVideo
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/runway_aleph/video_to_video"
        RESPONSE_CLASS = Types::TaskCreateResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedVideoToVideoResponse

        def initialize(http)
          @http = http
        end

        def run(**params)
          task = create(**params)
          poll_until_complete { get(task.id) }
        end

        def create(**params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params)
        end

        def get(id)
          request(:get, "#{ENDPOINT}/#{id}")
        end

        private

        def validate_params!(params)
          raise Core::ValidationError, "prompt is required" unless param(params, :prompt)
          raise Core::ValidationError, "video_url is required" unless param(params, :video_url)
          validate_optional!(params, :aspect_ratio, Types::ASPECT_RATIOS)
        end
      end
    end
  end
end

File: js/src/resources/video-to-video.ts
import type { HttpClient, PollingOptions, RequestOptions } from '@runapi.ai/core';
import { compactParams } from '@runapi.ai/core';
import { pollUntilComplete } from '@runapi.ai/core/internal';
import type { CompletedVideoToVideoResponse, TaskCreateResponse, VideoToVideoParams, VideoToVideoResponse } from '../types';

const ENDPOINT = '/api/v1/runway_aleph/video_to_video';

export class VideoToVideo {
  constructor(private readonly http: HttpClient) {}

  async run(params: VideoToVideoParams, options?: RequestOptions & PollingOptions): Promise<CompletedVideoToVideoResponse> {
    const { id } = await this.create(params, options);
    const response = await pollUntilComplete<VideoToVideoResponse>(() => this.get(id, options), {
      maxWaitMs: options?.maxWaitMs,
      pollIntervalMs: options?.pollIntervalMs,
    });
    return response as CompletedVideoToVideoResponse;
  }

  async create(params: VideoToVideoParams, options?: RequestOptions): Promise<TaskCreateResponse> {
    return this.http.request<TaskCreateResponse>('POST', ENDPOINT, { body: compactParams(params), ...options });
  }

  async get(id: string, options?: RequestOptions): Promise<VideoToVideoResponse> {
    return this.http.request<VideoToVideoResponse>('GET', `${ENDPOINT}/${id}`, options ?? {});
  }
}
